### PR TITLE
Added ability to set prompt parameter in OpenIdConnectOptions

### PR DESCRIPTION
**Pull request by gb056**
- Set prompt parameter to value set in Options in OpenIdConnectHandler.HandleChallengeAsync()
 - See http://openid.net/specs/openid-connect-core-1_0.html#AuthRequest for expected behavior

Addresses #1398

### CODE REVIEW

**Comment by gb056** [comment]
Default value of `OpenIdConnectOptions.Prompt` is hard coded to "none" for now, it would be great if we could use constants for this as with `ResponseType`

File: src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectOptions.cs [hunk 1]
```diff
@@ -210,6 +210,11 @@ public new OpenIdConnectEvents Events
         public string ResponseType { get; set; } = OpenIdConnectResponseType.IdToken;
 
         /// <summary>
+        /// Gets or sets the 'prompt'.
+        /// </summary>
+        public string Prompt { get; set; } = "none";
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Authentication.OpenIdConnect/OpenIdConnectOptions.cs]: Why a default rather than empty?

**Comment by PinpointTownes** [comment]
> Default value of OpenIdConnectOptions.Prompt is hard coded to "none" for now.

I agree with @Tratcher, you definitely can't do that. `prompt=none` asks the IdP to "silently" process the authorization request without displaying any consent page nor redirecting the user to a login page. If it can't do that, it must return an error to the client application indicating why the request failed.

It would be a terrible regression: folks who are not already logged in with the IdP or haven't granted your ASP.NET Core app an access to their resources wouldn't be able to successfully log in.

**Comment by gb056** [comment]
Of course, sorry clearly didn't read the spec properly there. Assumed `none` was the default without really thinking. Apologies, will fix now.

**Comment by poke** [comment]
I just noticed that this functionality is coming in 2.1. Is there any way we could extend this to allow overwriting the prompt value when challenging the authentication?

I image that we could simply look for an item in the `AuthenticationProperties` and use that if there’s one. Something like this:

```
Prompt = properties.Items.TryGetValue("prompt", out var prompt) ? prompt : Options.Prompt,
```

Then, one could explicitly require a login prompt like this:

```
var props = new AuthenticationProperties();
props.Items.Add("prompt", "login");
await HttpContext.ChallengeAsync(OpenIdConnectDefaults.AuthenticationScheme, props);
```

**Comment by Tratcher** [comment]
@poke see https://github.com/aspnet/Security/issues/990 for a related discussion

**Comment by poke** [comment]
@Tratcher Thanks, that’s interesting. But I’m not sure if a solution that’s being created for that will have an impact on OpenIdConnect. The OIDC handler is not built on top of the OAuth handler, so we would need a separate implementation anway. And since OIDC is limited by the capabilities of [`OpenIdConnectMessage`](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/blob/master/src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectMessage.cs), a dynamic way to build the query arguments on this side will probably not work (unless we also adjust the upstream message object).

I would happily provide a pull request for something like I’ve shown above, if there’s a chance for that being accepted. Ideally to get this done still in time for 2.1.0 (if that’s even possible).

**Comment by Tratcher** [comment]
I referenced it as conceptually similar, the Google Auth handler has been doing dynamic parameter overrides for a long time and we've considered expanding that to the other handlers.

Try a PR and we'll see how the pattern works out.

**Pull request merged**